src/type/defaultValues.js

```diff
@@ -0,0 +1,28 @@
+import { invariant } from '../jsutils/invariant';
+import type { ConstValueNode } from '../language/ast';
+
+import { astFromValue } from '../utilities/astFromValue';
+import { valueFromAST } from '../utilities/valueFromAST';
+
+import type { GraphQLInputType, GraphQLDefaultValueUsage } from './definition';
+
+export function getLiteralDefaultValue(
+  usage: GraphQLDefaultValueUsage,
+  type: GraphQLInputType,
+): ConstValueNode {
+  if (usage.literal) {
+    return usage.literal;
+  }
+  const literal = astFromValue(usage.value, type);
+  invariant(literal, 'Value cannot be converted to literal for this type');
+  return literal;
+}
+
+export function getCoercedDefaultValue(
+  usage: GraphQLDefaultValueUsage,
+  type: GraphQLInputType,
+): mixed {
+  return usage.value !== undefined
+    ? usage.value
```

Comment:
Looks to me like valueFromAST gets called during execution, could you call it at schema build time and just save the result as the defaultValue.value.

Just curious also, why do you save the default value node separately as part of the type if the entire ast is also preserved, why not just read from there?

Reply:
Calling `valueFromAST` during schema construction causes a circular dependency that can result in some TDZ issues. Also it's possible that the value is invalid, and schema validation occurs after schema construction, so we need to preserve the pre-coerced values.

We save the default value literal AST separately to simplify the logic - otherwise detecting a defaultValue would require inspecting the configuration as well as the SDL's AST - this moves it to a single source of truth. It also allows a default value literal to be provided outside the context of SDL-based construction which is at least useful for writing unit tests and may also be useful in future cases as well.

Reply:
I see your points, but this still does not seem right to me. Taking a different tack, if node/ast SDL is so separate, we should not have separate literal preserved at all and scalars that require uncoercion would have functions defined that allow that or default to serialize

i thought that your other related PRs were headed in that direction, but this seems to suggest not, and then makes ast values canonical for default values, which seems less optimal to me from a purist standpoint

i am not sure I am a purist, so, I could ask this the other way, why can’t we then preserve directive ast nodes on the type as well (or maybe just a subset of them that are marked in some way) one argument was the SDL / code separation.

Anyway, this seems overall inconsistent. I think a refactor with a staged build of types from ast in which default values are added last might help. If the default value is not valid, but you want to preserve the ast, you could save it in the defaultValue as an Error or have another way of storing it within the schema.

it just doesn’t make sense to me to delay until runtime something that can be calculated at schema construction time. It’s a sign that something is off imo

Reply:
Maybe a comment “todo: add memoization if/when possible” like elsewhere in code base. Thanks for the discussion, certainly been clarifying.

i meant mostly for schema, not query directives.

as an aside, not sure yet how these PRs affect schema directive default value processing generally

Reply:
Just a little unnecessary follow-up. I have spent new few hours trying to put my money where my mouth and implementing an alternative solution and was humbled quite quickly by circular references and preserving literals that error on coercion, and definitely now see the benefit of consolidating the literal with the default value separately from the AST considering how often the code has to check for the existence of either, i.e. all the points that @leebyron brought up.

Thanks @leebyron very much for indulging my comments and teaching me along the way about some of the goals of GraphQL, the reference implemationation, etc, etc.

There may be a better way of doing this, but it certainly is not as simple as I thought. The solution here now seems much more elegant than it did to me earlier. Huge apologies.